src/sentry/replays/query.py

```diff
@@ -168,6 +196,84 @@ def query_replays_dataset(
     return raw_snql_query(snuba_request, "replays.query.query_replays_dataset")
 
 
+def query_replays_dataset_with_subquery(
+    project_ids: List[str],
+    start: datetime,
+    end: datetime,
+    sorting: List[OrderBy],
+    fields: List[str],
+    pagination: Optional[Paginators],
+    tenant_ids: dict[str, Any] | None = None,
+):
+    # simply get the eligible replay_ids, only look at segment_0
+    # TODO: make sure timestamp end is current now.
+    subquery_snuba_request = Request(
+        dataset="replays",
+        app_id="replay-backend-web",
+        query=Query(
+            match=Entity("replays"),
+            select=[
+                Column("replay_id"),
+                Function(
+                    "identity", parameters=[Column("replay_start_timestamp")], alias="started_at"
+                ),
+            ],
+            where=[
+                Condition(Column("project_id"), Op.IN, project_ids),
+                Condition(Column("timestamp"), Op.LT, datetime.now()),
+                Condition(Column("timestamp"), Op.GTE, start),
+                Condition(Column("segment_id"), Op.EQ, 0),
+            ],
+            orderby=sorting,
+            granularity=Granularity(3600),
+            limit=Limit(pagination.limit),
+            offset=Offset(pagination.offset),
+        ),
+        tenant_ids=tenant_ids,
+    )
+
+    replay_ids_to_filter_results = raw_snql_query(
+        subquery_snuba_request, "replays.query.query_replays_dataset_subquery"
+    )
+
+    replay_ids_to_filter = [r["replay_id"] for r in replay_ids_to_filter_results["data"]]
+
+    # do the full query to get all aggregated fields
+    snuba_request = Request(
+        dataset="replays",
+        app_id="replay-backend-web",
+        query=Query(
+            match=Entity("replays"),
+            select=make_select_statement(fields, sorting, []),
+            where=[
+                Condition(Column("project_id"), Op.IN, project_ids),
+                Condition(Column("timestamp"), Op.LT, datetime.now()),
+                Condition(Column("timestamp"), Op.GTE, start),
+                Condition(Column("replay_id"), Op.IN, replay_ids_to_filter),
+            ],
+            having=[
+                # Must include the first sequence otherwise the replay is too old.
+                Condition(Function("min", parameters=[Column("segment_id")]), Op.EQ, 0),
+                # Make sure we're not too old.
+                Condition(Column("started_at"), Op.LT, end),
+                # TODO: talk about changing semantics default "end sort" to use started_at instead of finished at
+                # Require non-archived replays.
+                Condition(Column("isArchived"), Op.EQ, 0),
```

Comment:
How do you handle replay_ids that do not match the filters in the having clause?  If every replay returned in the first query is archived this will produce an empty result set.

Reply:
> Currently, if there are archived replays, this can break pagination as any rows in the first query that are filtered out by the second query will cause a reduced number of rows to be returned. I think this is acceptable for now as archival isn't an extremely common use-case, and pending our progress on a materialized view, we could look at solutions to this, for example: 

Oof, I missed this part.